### ISSUE TITLE
CANONICAL_REJECTED

### ISSUE THREAD

**Issue by AntonioAmore** [Norconex/collector-http#111]
I got following entries in log:

```
INFO -          DOCUMENT_FETCHED: http://www.site.com/data-policy
INFO -        REJECTED_CANONICAL: http://www.site.com/data-policy
INFO -           REJECTED_IMPORT: http://www.site.com/data-policy
```

Tell me please why is it rejected, is there appeared a new default configuration entry? Migrated from 2.0.1 to 2.2.0 Snapshot.

**Comment by essiembre** [comment]
This is a new feature introduced in 2.2.0 (from feature request #79) to help reduce crawling duplicate content.  You can read more about it [here](https://support.google.com/webmasters/answer/139066?hl=en).  The URL ``http://www.site.com/data-policy`` is rejected because in its HTML content, you have this:
```xml
<link rel="canonical" href="http://www.salesforce.com/platform/overview/"/>
```
The authors of that page are telling browsers the authoritative version is ``http://www.salesforce.com/platform/overview/`` and consequently that page should be crawled instead.

This means that canonical page will get crawled instead, unless you have filtering rules forcing to stay on the ``www.site.com`` domain.

You can disable this new feature by adding this tag to your crawler:
```xml
<canonicalLinkDetector ignore="true" />
```

**Comment by AntonioAmore** [comment]
Thanks a lot, this feature looks very helpful.

**Comment by essiembre** [comment]
That feature is now officially released.

**Comment by liar666** [comment]
Hi,

Sorry to reopen an old issue.

In my config file, I have:
```
  <startURLs stayOnDomain="true" stayOnPort="true" stayOnProtocol="true">
        <url>https://www.researchgate.net/publication/261340078_Introduction_of_sustainable_development_in_engineers%27_curricula_Problematic_and_evaluation_methods</url>
...
```

When I run the crawler, I get:

```
ResearchGate: 2016-11-28 15:50:31 INFO - ResearchGate: Crawling references...
ResearchGate: 2016-11-28 15:50:33 INFO -          DOCUMENT_FETCHED: https://www.researchgate.net/publication/261340078_Introduction_of_sustainable_development_in_engineers%27_curricula_Problematic_and_evaluation_methods
ResearchGate: 2016-11-28 15:50:33 INFO -        REJECTED_CANONICAL: https://www.researchgate.net/publication/261340078_Introduction_of_sustainable_development_in_engineers%27_curricula_Problematic_and_evaluation_methods
ResearchGate: 2016-11-28 15:50:42 INFO - ResearchGate: 100% completed (2 processed/2 total)
ResearchGate: 2016-11-28 15:50:42 INFO - ResearchGate: Reprocessing any cached/orphan references...
ResearchGate: 2016-11-28 15:50:42 INFO - ResearchGate: Crawler finishing: committing documents.
ResearchGate: 2016-11-28 15:50:42 INFO - ResearchGate: 2 reference(s) processed.
ResearchGate: 2016-11-28 15:50:42 INFO -          CRAWLER_FINISHED
```

When I have a look inside the website page code, I see:
` <link rel="canonical" href="https://www.researchgate.net/publication/281917092_Empreinte_ecologique_Comparer_la_demande_et_l'offre_de_ressources_regeneratives_de_la_biosphere" />`

This is exactly the start URL, but not encoded (' instead of %27).

Not sure if its a bug in the canonicalLinkDetector (which does not detect that links are the same) or it is in the website that forgot to encode the URL...

**Comment by essiembre** [comment]
It is both.  First there was indeed a bug with the GenericCanonicalLinkDetector.  It was truncating URLs with single quotes in them, thinking the single quote was the closing delimiter for the URL string.   This is now fixed in the latest [snapshot release](https://www.norconex.com/collectors/collector-http/download).  

Second, the single quote is a reserved characters (sub-delimiter) for URLs as per [RFC 3986](https://tools.ietf.org/html/rfc3986#page-12).  This means if you are keeping the default normalization rules, the rule ``encodeNonURICharacters`` has no effect on the quote and it is not escaped (rightfully so).  If the quote is not meant as a delimiter, it should always be encoded on the site.  You can get around this problem by forcing escaping of the single quote, like this:

```xml
      <urlNormalizer class="com.norconex.collector.http.url.impl.GenericURLNormalizer">
        <normalizations>
            removeFragment, lowerCaseSchemeHost, upperCaseEscapeSequence,
            decodeUnreservedCharacters, removeDefaultPort, 
            encodeNonURICharacters
        </normalizations>
        <replacements>
          <replace>
             <match>'</match>
             <replacement>%27</replacement>
          </replace>
        </replacements>
      </urlNormalizer>
```

Without that suggested normalization replacement, it will work with the fix, but the non-encoded canonical version will be followed instead of the encoded version.  Depends which one you want.  Please confirm.

**Comment by essiembre** [comment]
Fix is now in [2.6.1](https://www.norconex.com/collectors/collector-http/download).
